### PR TITLE
[REVIEW] Add `Index.set_names` api

### PR DESCRIPTION
**Pull request by galipremsagar**
Resolves: #6870 

This PR adds support for `set_names` API in both `Index` & `MultiIndex`.

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: CHANGELOG.md [hunk 1]
```diff
@@ -50,7 +52,6 @@
 - PR #6765 Cupy fallback for __array_function__ and __array_ufunc__ for cudf.Series
 - PR #6817 Add support for scatter() on lists-of-struct columns
 - PR #6805 Implement `cudf::detail::copy_if` for `decimal32` and `decimal64`
-- PR #6847 Add a cmake find module for cuFile in JNI code
```
Comment (galipremsagar) [inline CHANGELOG.md:53]: This PR was merged to `0.18` but entry is in the `0.17` section, so moved this around.

**Review by kkraus14 — CHANGES_REQUESTED** [review]

File: python/cudf/cudf/core/index.py [hunk 2]
```diff
@@ -452,6 +452,86 @@ def _clean_nulls_from_index(self):
...
+        """
+        Number of levels.
+        """
+        return 1
+
+    def _set_names(self, names, inplace=False):
+        if inplace:
+            idx = self
+        else:
+            idx = self.copy(deep=False)
```
[the remark this answers is not in the document]
Comment (galipremsagar) [reply python/cudf/cudf/core/index.py:466]: Pandas returns a shallow copy when `inplace=False`

File: python/cudf/cudf/core/multiindex.py [hunk 3]
```diff
@@ -190,6 +190,43 @@ def names(self, value):
...
+            raise ValueError("Length of names must match length of level.")
+        if level is None and len(names) != self.nlevels:
+            raise ValueError(
+                "Length of names must match number of levels in MultiIndex."
+            )
+
+        if self.nlevels > 1 and level is not None:
+            existing_names = list(self.names)
+            for i, l in enumerate(level):
+                existing_names[l] = names[i]
```
[the remark this answers is not in the document]
Comment (galipremsagar) [reply python/cudf/cudf/core/multiindex.py:225]: Ah, I missed this one. Adding it.

File: python/cudf/cudf/tests/test_multiindex.py [hunk 4]
```diff
@@ -1340,3 +1340,76 @@ def test_multiIndex_argsort(pdi, ascending):
...
+)
+@pytest.mark.parametrize(
+    "level, names",
+    [
+        (0, "abc"),
+        (1, "xyz"),
+        ([2, 1], ["a", "b"]),
+        (None, ["a", "b", "c"]),
+        (None, ["a", None, "c"]),
+    ],
```
[the remark this answers is not in the document]
Comment (galipremsagar) [reply python/cudf/cudf/tests/test_multiindex.py:1385]: Done

File: python/cudf/cudf/core/multiindex.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (galipremsagar) [reply python/cudf/cudf/core/multiindex.py]: Handled this 👍

File: python/cudf/cudf/tests/test_multiindex.py [hunk 5]
```diff
@@ -1340,3 +1340,110 @@ def test_multiIndex_argsort(pdi, ascending):
...
+)
+@pytest.mark.parametrize(
+    "level, names",
+    [
+        ([None], "abc"),
+        (["three", "one"], ["a", "b"]),
+        (["three", 1], ["a", "b"]),
+        ([0, "three", 1], ["a", "b", "z"]),
+        (["one", 1, "three"], ["a", "b", "z"]),
+        (["one", None, "three"], ["a", "b", "z"]),
```
Comment (kkraus14) [inline python/cudf/cudf/tests/test_multiindex.py:1418]: I hate to ask this... but what's the behavior is someone passes in say `[0, 1]` and the multiindex levels are named `[1, 0]`? Does Pandas resolve to using integers as positions or names first? Should we test this?

**Review by kkraus14 — APPROVED** [review]

File: python/cudf/cudf/tests/test_multiindex.py (hunk 5 above)
Comment (galipremsagar) [reply python/cudf/cudf/tests/test_multiindex.py:1418]: > Does Pandas resolve to using integers as positions or names first? 

Pandas tries to resolve the input first as labels and then as positions. Similar to what our `MultiIndex._level_index_from_level` does.

> Should we test this?

Yep, added tests for this in `test_multiindex_set_names_default_and_int_names`

**Pull request merged**